### PR TITLE
Build objects

### PR DESCRIPTION
**Pull request by vloginova**
- Rgraph improvements and refactoring
- Build object files for a modular application

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="http://app.review.ninja/abusalimov/mybuild/pull/4" target="_blank">ReviewNinja

File: glue.py [hunk 1]
```diff
@@ -47,11 +49,19 @@ def options(self, module, ctx):
     def configure(self, module, ctx):
         ctx.load(self.waf_tools)
 
+def interpolate_string(string, env):
```
Comment (abusalimov) [inline glue.py]: - [x] This could be implemented through `string.Template(s).substitute(env)`, without messing with regexes.

File: glue.py [hunk 2]
```diff
@@ -64,10 +74,27 @@ def define(self, key, val):
         self.build_kwargs['defines'].append('{0}={1}'.format(key, val))
 
     def build(self, module, ctx):
-        self.build_kwargs['use'] = [m._name for m in module.depends]
-        self.build_kwargs['source'] = module.files
+        sources = []
+        objects = []
+
+        for fname in module.files:
+            if re.match('.*\.o', fname):
```
Comment (abusalimov) [inline glue.py]: - [x] Missing end anchor (otherwise `foo.oww` matches as well). Why not just `fname.endswith('.o')`?

File: glue.py (hunk 1 above)
Comment (abusalimov) [inline glue.py]: Plus I would probably move it to `util.misc` package. Or not.

File: glue.py [hunk 3]
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
...
-        ctx.objects(**self.build_kwargs)
+        ctx(features='c', **self.build_kwargs)
 
 
 class CcAppTool(CcTool):
     def build(self, module, ctx):
         super(CcAppTool, self).build(module, ctx)
-        ctx.program(**self.build_kwargs)
+
+        use = [instance._name for instance in ctx.instance_map \
```
Comment (abusalimov) [inline glue.py]: - [x] Avoid line continuation (from [PEP 8](https://www.python.org/dev/peps/pep-0008/#id14)):

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

File: glue.py [hunk 4]
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
...
+
+class GenHeadersTool(WafBasedTool):
+    def get_headers(self, module):
+        headers = []
+
+        project_relative_path = '/'.join(module.__module__.split('.')[1:-1])
+        preproc_relative_path = '../../src/{0}/'.format(project_relative_path)
+
+        for fname in module.files:
+            if re.match('.*\.h', fname):
```
Comment (abusalimov) [inline glue.py]: - [x] `fname.endswith(...)`

File: glue.py [hunk 5]
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
...
+                headers.append(preproc_relative_path + fname)
+        return headers
+
+    def get_option_string(self, mod_name, opt_name, value):
+        if isinstance(value, str):
+            opt_type = 'STRING'
+        elif isinstance(value, bool):
+            opt_type = 'BOOLEAN'
+            value = 1 if value else 0
```
Comment (abusalimov) [inline glue.py]: - [x] `value = int(value)`

File: glue.py [hunk 6]
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
...
+        elif isinstance(value, int):
+            opt_type = 'NUMBER'
         else:
-            ctx.shlib(**self.build_kwargs)
+            raise Exception('Option with type {0} is not supported' \
+                            .format(type(value)))
+
+        fmt = 'OPTION_{TYPE}_{MOD}__{NAME} {VALUE}'
```
Comment (abusalimov) [inline glue.py]: - [x] Or `return 'OPTION_{opt_type}_{mod_name}__{opt_name} {value}'.format(**locals())`

File: glue.py [hunk 7]
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
...
+
+        return fmt.format(TYPE=opt_type, MOD=mod_name, NAME=opt_name,
+                          VALUE=value)
+
+    def get_options(self, module):
+        options = []
+
+        options_dict = module._ModuleBase__optuple.__dict__
```
Comment (abusalimov) [inline glue.py]: - [x] Do not ever access private fields!

File: glue.py [hunk 8]
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
...
+        if isinstance(value, str):
+            opt_type = 'STRING'
+        elif isinstance(value, bool):
+            opt_type = 'BOOLEAN'
+            value = 1 if value else 0
+        elif isinstance(value, int):
+            opt_type = 'NUMBER'
         else:
-            ctx.shlib(**self.build_kwargs)
+            raise Exception('Option with type {0} is not supported' \
```
Comment (abusalimov) [inline glue.py]: - [x] Unnecessary line continuation.

File: glue.py [hunk 9]
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
...
+        fmt = 'OPTION_{TYPE}_{MOD}__{NAME} {VALUE}'
+
+        return fmt.format(TYPE=opt_type, MOD=mod_name, NAME=opt_name,
+                          VALUE=value)
+
+    def get_options(self, module):
+        options = []
+
+        options_dict = module._ModuleBase__optuple.__dict__
+        for opt_name, opt_val in iteritems(options_dict):
```
Comment (abusalimov) [inline glue.py]: - [x] Try this one:

```
for option, value in module._optuple._iterpairs():
    ...
```

**Comment by abusalimov** [comment]
I'm going to fix these notes above and merge, although I didn't review rgraph part carefully.

**Comment by vloginova** [comment]
Ok, thank you. You can review rgraph later and I'll happily fix it.

**Pull request merged**